src/model.rs

```diff
@@ -121,23 +124,24 @@ impl Analysis<Mdl> for TensorAnalysis {
         let x = |i: &Id| &egraph[*i].data;
         let mut g = egraph.analysis.graph.borrow_mut();
         match enode {
-            Mdl::Matmul([a, b]) => {
+            Mdl::Matmul([act, a, b]) => {
                 // Check types
+                assert!(x(act).dtype == DataKind::Scalar);
                 assert!(x(a).dtype == DataKind::Tnsr);
                 assert!(x(b).dtype == DataKind::Tnsr);
 
                 // Get arguments
                 let t_a = x(a).meta;
                 let t_b = x(b).meta;
+                let activation: ActiMode = x(act).val.try_into().unwrap();
 
                 // Create tensorhandle and get metadata
                 unsafe {
-                    let mm = g.matmul(t_a, t_b, ActiMode_AC_MODE_NONE);
-                    //let r_cost = (*(*mm).op.ptr).runtime;
-                    //println!("Cost of matmul is {}", r_cost);
+                    let mm = g.matmul(t_a, t_b, activation);
                     Self::Data {
                         dtype: DataKind::Tnsr,
                         val: 0,
+                        name: String::new(),
                         meta: mm,
                     }
```

Comment:
Only the call to `g.matmul` is unsafe. Move the struct definition outside of the `unsafe` block.
```
let mm = unsafe { g.matmul(t_a, t_b, activation) };
Self::Data {
  dtype: DataKind::Tnsr,
  val: 0,
  name: String::new(),
  meta: mm,
}
```

In general, try to keep unsafe blocks as small as possible. For reviewers, `unsafe` is a flag that we should take additional time checking for bad behaviors. By keeping the block as small as possible, you reduce the scope we have to check.

Reply:
Please also do this for subsequent occurrences of `unsafe`.

Reply:
Thanks for this tip! Updated all unsafe